GPII-3381: Added basic script to load empty prefs sets into CouchDB

javihernandez opened the pull request:
Hey there. This is a first approach to load empty prefs sets into the preferences server, which is only meant to be used by DevOps to create empty prefs sets into the preferences server.

cindyli commented on scripts/loadEmptyPrefsSets.js:
```diff
@@ -0,0 +1,127 @@
+
+gpii.uuidLoader.generatePrefsSet = function () {
+    var gpiiKey = uuid.v4();
+    var currentTime = new Date().toISOString();
```
The logic from line 57-83 is similar to [convertPrefs.js](https://github.com/GPII/universal/blob/master/scripts/convertPrefs.js#L43-L70). Can you create a shared function for it so the data structure generation only needs to be maintained at one place? Thanks.

cindyli commented on scripts/loadEmptyPrefsSets.js:
```diff
@@ -0,0 +1,127 @@
+
+require("gpii-pouchdb");
+
+var emptyPrefsSetBlock = {
```
This variable can be defined in gpii.uuidLoader.generatePrefsSet().

cindyli commented on scripts/loadEmptyPrefsSets.js:
```diff
@@ -0,0 +1,127 @@
+    gradeNames: ["gpii.pouch"],
+    dbOptions: {
+        //name: "http://localhost:5984/gpii"
+        name: process.env.COUCHDB_URL
```
Use a meaningful option name such as "dbUrl".

javihernandez replied:
I'm just using the "gpii.pouch" component's dbOptions, https://github.com/GPII/gpii-pouchdb/blob/master/docs/pouchdb.md#component-options.

mrtyler commented:
Since this script is not idempotent, how shall we run it in long-lived environments like production?

javihernandez commented:
@cindyli, I've just pushed a few commits to address your comments. I've also updated the convertPrefs.js script to use the shared function. Let me know what you think about it.

@mrtyler, probably there is some misunderstanding on my side, but from our last conversations, I think the decision was that this script was only meant to create empty and uuid-based prefsSets. And it's true that as I'm implementing it, I think we should consider extending it to be more flexible (or even create another script/s), but I'm just sticking to the "plan" for this one. Anyways, we can chat about it anytime.

Oh, and I almost forgot, I've added a check to ensure that the document we're about to push doesn't exist in DB yet, but at this moment the script will stop running when there is an error like this one. This is, in fact, the trigger for these questions:
* What do we do when a prefsSet already exists in DB?
** Do we stop the process?
** Do we overwrite it?
** Do we skip this one and create another UUID and complete the "n" additions?
* What to do when there is a momentary connection-related issue?
** Do we stop the process?
** Do we try keep going (by adding a timeout just in case it's really not gonna happen)?

Do we eventually need to check with UX the life cycle of a prefsSet?

cindyli commented on scripts/prefsSetsDbUtils.js:
```diff
@@ -0,0 +1,69 @@
+/*!
```
Can we create a "script/shared" directory and move this script there? Also, add a comment in the script about who's sharing it and the goal of this sript.

mrtyler commented:
@javihernandez 
> this script was only meant to create empty and uuid-based prefsSets

I agree, and that's fine. But I'm not sure how this answers my question about how the script will be run against long-lived environments like production. Asked another way: how will empty preference sets be created in production?

javihernandez commented:
> I agree, and that's fine. But I'm not sure how this answers my question about how the script will be run against long-lived environments like production. Asked another way: how will empty preference sets be created in production?

Ok, the script can be called this way:
```COUCHDB_URL="http://localhost:8059/gpii" NEW_USERS=20 node scripts/loadUniquePrefsSets.js```

This will create 20 empty users with a uuid that will be generated on the fly during the execution of the script. 
The script renders the new users as they are inserted in the DB (I can just dump them in a file at the end of the process if we want to). 

To your question, "how can this be executed?" 
First, this script is not meant to be executed as part of a pipeline or periodically, but manually and on-demand, and only until we get to the point where users can create their own prefsSets.
For this reason, I think that we can follow a similar approach to the one described [here](https://github.com/gpii-ops/gpii-infra/blob/master/aws/README.md#manual-process-generating-db-data-of-user-preferences-and-load-it-into-a-clusters-preferences-server). This is, by opening a port forwarding between the cluster's couchdb and your local machine to run the script against.

However, I'd like to know your thoughts on this, is there any alternative that you can think of?

mrtyler commented:
@javihernandez 
> I've added a check to ensure that the document we're about to push doesn't exist in DB yet, but at this moment the script will stop running when there is an error like this one. This is, in fact, the trigger for these questions:
> * What do we do when a prefsSet already exists in DB?
** Do we stop the process?
** Do we overwrite it?
** Do we skip this one and create another UUID and complete the "n" additions?

Unless I'm missing some Infusion magic, a UUID is "guaranteed" to be universally unique so this situation would be pretty surprising. I think bailing out immediately -- and letting the user know how surprising this situation is -- is sensible. 

> * What to do when there is a momentary connection-related issue?
** Do we stop the process?
** Do we try keep going (by adding a timeout just in case it's really not gonna happen)?

Since this script is designed only for manual, interactive use, I think I would do whatever is easiest to implement.

If my machine has connection problems halfway through and I have to run it again with `NEW_USERS=10` instead of `NEW_USERS=20`, I don't think that's a big deal (especially when creating 30 new users instead of 20 new users has negligible impact on the universe). Similarly, I'm not worried about an interactive process that hangs forever (though I suspect that Infusion or whatever HTTP library has a default timeout somewhere to help prevent such a problem).

mrtyler commented on scripts/loadEmptyPrefsSets.js:
```diff
@@ -0,0 +1,103 @@
+function createNewUsers(uuidLoader) {
+    var data = gpii.prefsSetsDbUtils.generatePrefsSet(uuid.v4(), gpii.prefsSetsDbUtils.emptyPrefsSetBlock);
+    uuidLoader.addPrefsSet(data.key, data.prefsSafe).then(function (newUser) {
+        console.log("## New user created:", newUser);
```
Given our discussion about error conditions, I think it might be helpful to include the value of `n` here, as in "User #n created".

If you're feeling fancy, save the original `n` so you can say things like "User 5 of 20 created". :)

cindyli commented:
@mrtyler's comments remind me it would be better to write the usage of the upload script as a top comment in the script. 

Also, this script probably should generate a .csv or .txt file that has all the generated keys. I guess the person who requested the creation of these keys would love this file. What do you think to have the upload script take in a parameter of where the output file should be written?

javihernandez commented:
@cindyli and @mrtyler this is ready again for another round of review ;)

cindyli commented on scripts/loadEmptyPrefsSets.js:
```diff
@@ -0,0 +1,127 @@
+//   * COUCHDB_URL: This neeeds to be in the form "http://host/dbname", but in case
+//   you need to provide the credentials to CouchDB, you need to use the form
+//   "http://user:pass@host:port/dbname".
+//   * NEW_USERS: The amount of new empty preferences sets you want to create
```
Might be less confusing to use "keys" rather than "users" in this script.

Can you also mark each option with "required" or "optional". Thanks.

javihernandez replied:
> Might be less confusing to use "keys" rather than "users" in this script.

Okay, got a question about terminology. What I'm calling here "user" is represented by 2 tiers, a "prefSafe" and a "gpiiKey", right?. Is "key" the right term to use when we are referring to this? Sorry, I just want to understand and use the right terminology.

cindyli replied:
Good point. Now I'm also not sure which term should be used. ;) @amb26, what do you think?

cindyli commented on scripts/loadEmptyPrefsSets.js:
```diff
@@ -0,0 +1,127 @@
+
+fluid.defaults("gpii.uuidLoader", {
+    gradeNames: ["gpii.pouch"],
+    outputFileName: process.env.OUTPUT_FILE || "generated-prefSets-" + new Date().toISOString(),
```
Shall a file extension "txt" or "csv" be added to the default output file name so that it can be auto opened at double click?

javihernandez replied:
Good point!

cindyli commented on scripts/loadEmptyPrefsSets.js:
```diff
@@ -0,0 +1,127 @@
+
+// The real action
+//
+function createNewUsers(uuidLoader) {
```
Namespace this function too.

javihernandez replied:
Do you also want me to make it part of the component and make it available as an invoker?

cindyli replied:
I think it's great to have the whole process in fluid components.

cindyli commented on scripts/loadEmptyPrefsSets.js:
```diff
@@ -0,0 +1,127 @@
+        }
+    }, function (error) {
+        console.log("Got an error, let's stop it here. :/", error);
+        console.log("You can find the users we could create before crashing here:", uuidLoader.options.outputFileName);
```
"could create" -> "have created"

cindyli commented on scripts/loadEmptyPrefsSets.js:
```diff
@@ -0,0 +1,127 @@
+            createNewUsers(uuidLoader);
+        } else {
+            console.log("We're done adding new users! :)");
+            console.log("Remember, you have a list of the created users here:", uuidLoader.options.outputFileName);
```
Nice reminder. :)

cindyli commented on scripts/loadEmptyPrefsSets.js:
```diff
@@ -0,0 +1,127 @@
+
+// This triggers the "real action"
+//
+var n = parseInt(process.env.NEW_USERS);
```
Namespace this variable too. All appearances of `process.env.NEW_USERS` at line 101, 102 can be replaced by this variable.

cindyli replied:
Ah, i ses, you use `n` as a counter. I agree with @mrtyler, it's probably more readable to have 2 variables to track different things: 

```
gpii.loader.totalNumOfKeys = parseInt(process.env.NEW_USERS);
gpii.loader.count = 0;
```

cindyli commented on scripts/loadEmptyPrefsSets.js:
```diff
@@ -0,0 +1,127 @@
+//   * COUCHDB_URL: This neeeds to be in the form "http://host/dbname", but in case
+//   you need to provide the credentials to CouchDB, you need to use the form
+//   "http://user:pass@host:port/dbname".
+//   * NEW_USERS: The amount of new empty preferences sets you want to create
```
Rename `NEW_USERS` to `NUM_OF_KEYS`.

javihernandez commented on scripts/loadEmptyPrefsSets.js:
```diff
@@ -82,22 +95,33 @@ gpii.uuidLoader.addPrefsSet = function (that, gpiiKey, prefSafe) {
+        fs.appendFileSync(uuidLoader.options.outputFileName, newUser + "\n");
+        if (process.env.SHOW_PROGRESS) {
+            console.log("User",
+                        process.env.NEW_USERS - n + 1, "of",
```
That's absolutely correct @mrtyler, it's horrible! I'll follow Cindy's advice below and this will go away.

javihernandez commented:
This is ready again for review

cindyli commented on scripts/loadEmptyKeys.js:
```diff
@@ -0,0 +1,132 @@
+// This triggers the "real action"
+//
+var uuidLoader = gpii.uuidLoader();
+if (uuidLoader.options.totalNumOfKeys > 0) {
```
Output a message in the `else` block.

Output an error when required field (`uuidLoader.options.totalNumOfKeys` or `uuidLoader.options.dbOptions.name`) are not provided.

javihernandez replied:
The only thing I can do is to check that COUCHDB_URL is empty before instantiating my component, this is because the grade "gpii.pouch" will raise an error if a wrong dbOptions.name is provided. gpii.pouch doesn't provide me a way to catch that error and provide a proper message. Any advice?

cindyli replied:
If so, checking COUCHDB_URL may not be necessary if "gpii.pouch" will raise and error and terminate the process if an empty or incorrect COUCHDB_URL is provided.

cindyli commented on scripts/loadEmptyKeys.js:
```diff
@@ -0,0 +1,132 @@
+//
+var uuidLoader = gpii.uuidLoader();
+if (uuidLoader.options.totalNumOfKeys > 0) {
+    gpii.uuidLoader.count = 1;
```
Rather than using a global variable, `count` can be tracked as a member option of `uuidLoader`.

cindyli commented on scripts/loadEmptyKeys.js:
```diff
@@ -0,0 +1,132 @@
+        promise.reject({isError: true, message: "doc with id " + doc._id + " already exists in DB"});
+    }, function (err) {
+        if (err.message === "missing") {
+            that.put(doc).then(function (newDoc) {promise.resolve(newDoc);}, function (err) {promise.reject(err);});
```
Can you use a more radalbe format like what you've done for other promise callbacks. Thanks.

cindyli commented on scripts/shared/prefsSetsDbUtils.js:
```diff
@@ -0,0 +1,80 @@
+ * The data contains both the gpiiKey and the preferences safe.
+ *
+ * @param  {String} gpiiKeyId - The identifier of the prefsSet (usually, a uuid).
+ * @param  {Object} preferences - The preferences to include in the prefsSet.
```
This `preferences` block can be made optional. If not provided, use `gpii.prefsSetsDbUtils.emptyPreferencesBlock` as the default. Add this line into the function:
```
var preferences = preferences || gpii.prefsSetsDbUtils.emptyPreferencesBlock;
```
So that the caller script doesn't need to be aware of the existence of `gpii.prefsSetsDbUtils.emptyPreferencesBlock`.

javihernandez commented:
Hey @cindyli, this is ready again for another round of review

cindyli commented on scripts/loadEmptyKeys.js:
```diff
@@ -0,0 +1,137 @@
+    outputFileName: process.env.OUTPUT_FILENAME || "generated-keys-" + new Date().toISOString() + ".txt",
+    totalNumOfKeys: parseInt(process.env.NUM_OF_KEYS),
+    hideProgress: process.env.HIDE_PROGRESS,
+    count: 1,
```
`count` is not intended to be configurable but more for internal tracking, so it might make more sense to have it as a member option.

cindyli commented on scripts/loadEmptyKeys.js:
```diff
@@ -0,0 +1,137 @@
+    that.addKeyInDb(keyData).then(function (newKey) {
+        fs.appendFileSync(outputFileName, newKey + "\n");
+        if (!hideProgress) {
+            console.log("Key", count, "of", totalNumOfKeys, "created:", newKey);
```
Add spaces in between words for readability.

javihernandez replied:
@cindyli,

when we run the script you get the output like this: "Key 2 of 20 created: bdd229c3-e41e-4129-b632-de22bfd381e5"

Is this wrong? ... OR do you mean in code by appending the strings with "+" rather than with ","?

cindyli replied:
This message looks great. Nothing needs to be changed. Sorry about the comment.

javihernandez replied:
Okay, no prob, just wanted to double check.
